ops/pebble.py

```diff
@@ -752,5 +872,214 @@ def get_services(self, names: List[str] = None) -> List[ServiceInfo]:
+            name = part.get_param('name', header='Content-Disposition')
+            if name == 'response':
+                resp = _json_loads(part.get_payload())
+            elif name == 'files':
+                filename = part.get_filename()
+                if filename != path:
+                    raise ProtocolError('path not expected: {}'.format(filename))
+                # decode=True, ironically, avoids decoding bytes to str
+                content = part.get_payload(decode=True)
+
+        if resp is None:
+            raise ProtocolError('no "response" field in multipart body')
+        self._raise_on_path_error(resp, path)
+
+        if content is None:
+            raise ProtocolError('no file content in multipart response')
+        if encoding is not None:
+            content = content.decode(encoding)
+        return content
+
+    @staticmethod
+    def _raise_on_path_error(resp, path):
+        result = resp['result'] or []  # in case it's null instead of []
+        paths = {item['path']: item for item in result}
+        if path not in paths:
+            raise ProtocolError('path not found in response metadata: {}'.format(resp))
+        error = paths[path].get('error')
+        if error:
+            raise PathError(error['kind'], error['message'])
+
+    def write_content(
+            self, path: str, content: typing.Union[bytes, str], encoding: str = 'utf-8',
+            make_dirs: bool = False, permissions: int = None, user: typing.Union[str, int] = None,
+            group: typing.Union[str, int] = None):
+        """Write content to a given file path on the remote system.
+
+        Args:
+            path: Path of the file to write to on the remote system.
+            content: Content to write (str or bytes).
+            encoding: Encoding to use for encoding content str to bytes.
+                Ignored if content is bytes.
+            make_dirs: If True, create parent directories if they don't exist.
+            permissions: Permissions (mode) to create file with (Pebble default
+                is 0o644).
+            user: UID or username for file.
+            group: GID or group name for file.
+        """
+        info = self._make_auth_dict(permissions, user, group)
+        info['path'] = path
+        if make_dirs:
+            info['make-dirs'] = True
+        metadata = {
+            'action': 'write',
+            'files': [info],
+        }
+
+        multipart = MIMEMultipart('form-data')
+
+        part = MIMEBase('application', 'json')
+        part.add_header('Content-Disposition', 'form-data', name='request')
+        part.set_payload(json.dumps(metadata))
+        multipart.attach(part)
+
+        part = MIMEBase('application', 'octet-stream')
+        part.add_header('Content-Disposition', 'form-data', name='files', filename=path)
+        if isinstance(content, str):
+            content = content.encode(encoding)
+        part.set_payload(content)
+        multipart.attach(part)
```

Comment:
Same question as above on the opposite end.

Reply:
See above.